data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/stream/ShardConsumer.java

```diff
@@ -298,6 +319,34 @@ private void waitForExport() {
         }
     }
 
+    /**
+     * Only to skip processing when below two conditions are met.
+     * - Last Shard Iterator is provided (Shard with ending sequence number)
+     * - Last Event Timestamp is later than start time or No Last Event Timestamp (empty shard)
+     */
+    private boolean shouldSkip() {
+        // Do skip check
+        if (lastShardIterator != null && !lastShardIterator.isEmpty()) {
+            GetRecordsResponse response = callGetRecords(lastShardIterator);
```

Comment:
It appears that each iteration now requires two requests to `GetRecordsRequest` - one for the last shard iterator and another for the `shardIterator`. Is there a better way to avoid making both of these requests?

Reply:
A shard iterator is used to get changes event (via GetRecords api) from stream, now to prevent missing shards, we will also process closed shards (shard with ending sequence number), only for closed shard, we will check the last record in the shard and see if the event time is after than the export time, if not, simply completed the shard. Otherwise, just use the normal shard iterator to get records for processing.

Reply:
I think the question is why we call `callGetRecords` in the `shouldSkip` check, and then call it again right after. Why not call it one time and reuse that response to check the skip and then you don't need to call it again on line 222

Reply:
That was actually I did at the beginning,  passed something like `lastShardIterator == null? shardIterator: lastShardIterator` to the getRecords request, then set `lastShardIterator = null` if the should process to go to the normal process.  But it doesn't look tidy to me though there is only one GetRecords call in the code, it's just to check `if lastShardIterator == null` everywhere and differnet logic applies.   So I completely add a separate section for lastShardIterator which can be taken out in the future, there is another option that we don't need to process closed shard probably is something need to try in the future.